feat: update view properties after animation

sgtcoolguy opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-13764

**Description:**
After running `Ti.UI.View#animate()`, the animated properties should reflect their new values. This PR attempts to do just that.

Just prior to firing the `complete` event and calling the optional `callback` function, we apply the properties held by the animation to the underlying `Ti.UI.View`.

**Note:** This is a behavioral change! I would argue it is not a breaking change (unless I broke the implementation here somehow) - it's actually a sort of bug fix of "bad" behavior we've had forever.

I've updated a handful of tests we had that did animations and checked the end values - updating them to verify we use the target value from the animation. I *did not* try transition animations (one view to another).

**Another Note:** This was triggered by reviewing #11658 - where a `Ti.UI.Label`'s `color` property was animated. Note that iOS has an issue in animating a label color where it just immediately changes color rather than animating over time. It's an OS limitation. Annoying.

sgtcoolguy committed:
feat(android): update view properties post-animation

Fixes TIMOB-13764

sgtcoolguy committed:
test: update tests to handle post-animation values

sgtcoolguy committed:
feat(ios): update view properties post-animation

Fixes TIMOB-13764

sgtcoolguy committed:
test: validate post-animation values on ios

sgtcoolguy committed:
docs: add note about behavior change for animations

build commented:
<!--
  1 failure:  Tests have failed...
  1 warning:  tests/Resources/t...
  3 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  **tests/Resources/ti.ui.label.addontest.js#L11** - tests/Resources/ti.ui.label.addontest.js line 11 – 'utilities' is assigned a value but never used. (no-unused-vars)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11666/1/artifact/dist/mobilesdk-9.1.0.v20200428125742-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 2 tests have failed
There are 2 tests failing and 702 skipped out of 7346 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.UI.View</td><td>convertPointToView (9)</td><td>0.177</td><td><pre>Error: expected 223 to equal 123
at Assertion.fail (/node_modules/should/cjs/should.js:275:13)
      at Assertion.value (/node_modules/should/cjs/should.js:356:9)
      at View.listener (/ti.ui.view.test.js:649:18)
      at View.value (ti:/events.js:50:21)
      at View.value (ti:/events.js:102:19)</pre></td></tr>
<tr><td>ios.ipad.Titanium.UI.WebView</td><td>progress event (13.2.2)</td><td>1.151</td><td><pre>Error: expected 'https://www.google.com/#spf=1588104653651' to be equals one of Array [ 'https://www.google.com/', 'https://www.google.com' ]
fail@file:///Users/build/Library/Developer/CoreSimulator/Devices/480970E5-608F-4ABF-9DFE-32A4D2DCE2C8/data/Containers/Bundle/Application/397837DD-F729-4049-9E52-9CB7CC8D0B98/mocha.app/node_modules/should/cjs/should.js:275:19
value@file:///Users/build/Library/Developer/CoreSimulator/Devices/480970E5-608F-4ABF-9DFE-32A4D2DCE2C8/data/Containers/Bundle/Application/397837DD-F729-4049-9E52-9CB7CC8D0B98/mocha.app/node_modules/should/cjs/should.js:356:13
file:///Users/build/Library/Developer/CoreSimulator/Devices/480970E5-608F-4ABF-9DFE-32A4D2DCE2C8/data/Containers/Bundle/Application/397837DD-F729-4049-9E52-9CB7CC8D0B98/mocha.app/ti.ui.webview.test.js:733:28</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e5d6cc3b537d07e986282d8a6e06b571218db72f
</p>

vijaysingh-axway commented on apidoc/Titanium/UI/Animation.yml:
```diff
@@ -21,10 +21,12 @@ description: |
     `top`, `left`, `width`, `height`) are not changed by the animation. See the description of the
     [animate](Titanium.UI.View.animate) method for more information.
 
+    As of 9.1.0, the animated properties shoudl reflect their final values just before the `complete` event and/or the <Ti.UI.View.animate> callback is fired.
```
```suggestion
    As of 9.1.0, the animated properties should reflect their final values just before the `complete` event and/or the <Ti.UI.View.animate> callback is fired.
```

vijaysingh-axway reviewed (approved):
iOS changes looks good.

jquick-axway commented:
@sgtcoolguy, this PR crashes on iOS at the end of the animation if you set "autoreverse" to `true`.

```javascript
var window = Ti.UI.createWindow({
	backgroundColor: "white",
	fullscreen: true,
});
var squareView = Ti.UI.createView({
	backgroundColor: "blue",
	borderColor: "black",
	borderWidth: "2dp",
	width: "80dp",
	height: "80dp",
});
window.add(squareView);
window.addEventListener("postlayout", function onLayout() {
	window.removeEventListener("postlayout", onLayout);
	var windowSize = window.size;
	var minDimension = Math.min(windowSize.width, windowSize.height);
	var letterboxScale = minDimension / squareView.size.width;
	var animation = Ti.UI.createAnimation({
		transform: Ti.UI.createMatrix2D().scale(letterboxScale, letterboxScale),
		curve: Ti.UI.ANIMATION_CURVE_EASE_IN,
		duration: 2000,
		autoreverse: true,
//		repeat: 2,
	});
	squareView.animate(animation);
});
window.open();
```

Here is the stack trace...
```
[WARN]  New layout set while view [object TiUIView] animating: Will relayout after animation.
[WARN]  *** Terminating app due to uncaught exception 'org.appcelerator.TiUIViewProxy', reason: 'Invalid type passed to function'
[WARN]  *** First throw call stack:
[WARN]  (
[WARN]          0   CoreFoundation                      0x00000001081ca8db __exceptionPreprocess + 331
[WARN]          1   libobjc.A.dylib                     0x0000000106fd7ac5 objc_exception_throw + 48
[WARN]          2   TitaniumKit                         0x0000000106ea6574 JavascriptNameForClass + 0
[WARN]          3   TitaniumKit                         0x0000000106efb442 -[TiProxy applyProperties:] + 258
[WARN]          4   TitaniumKit                         0x0000000106ee7fe3 -[TiAnimation animationCompleted:finished:context:] + 327
[WARN]          5   UIKitCore                           0x00000001145b53e3 -[UIViewAnimationBlockDelegate _didEndBlockAnimation:finished:context:] + 847
[WARN]          6   UIKitCore                           0x0000000114588e47 -[UIViewAnimationState sendDelegateAnimationDidStop:finished:] + 343
[WARN]          7   UIKitCore                           0x0000000114589469 -[UIViewAnimationState animationDidStop:finished:] + 293
[WARN]          8   QuartzCore                          0x000000010595bc5b _ZN2CA5Layer23run_animation_callbacksEPv + 321
[WARN]          9   libdispatch.dylib                   0x000000010a600d02 _dispatch_client_callout + 8
[WARN]          10  libdispatch.dylib                   0x000000010a60ca50 _dispatch_main_queue_callback_4CF + 1276
[WARN]          11  CoreFoundation                      0x0000000108131a79 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
[WARN]          12  CoreFoundation                      0x000000010812c126 __CFRunLoopRun + 2310
[WARN]          13  CoreFoundation                      0x000000010812b4d2 CFRunLoopRunSpecific + 626
[WARN]          14  GraphicsServices                    0x000000010c70c2fe GSEventRunModal + 65
[WARN]          15  UIKitCore                           0x00000001140e2fc2 UIApplicationMain + 140
[WARN]          16  ClassicAppTest                      0x0000000101b1902c main + 1228
[WARN]          17  libdyld.dylib                       0x000000010a66e541 start + 1
[WARN]  )
[WARN]  libc++abi.dylib: terminating with uncaught exception of type NSException
```

jquick-axway reviewed (changes requested):
I don't think we should update the properties if "autoreverse" property is set to `true`, because the view will return back to its original settings at the end of the animation.

But other than that, yes, I agree with this change.

sgtcoolguy committed:
fix(ios): avoid applying props on autoreverse anims

sgtcoolguy committed:
test: add test for autoreverse anim

sgtcoolguy committed:
fix(android): don't apply end properties to autoreversed animations

sgtcoolguy commented:
ok pushed changes to guard against the crash @jquick-axway found on iOS, avoid applying the properties if the animation is autoreversed, and added a unit test for an autoreversed animation (which did surface the crash before I fixed it locally).

jquick-axway commented:
I tested the newest updates to this PR with the below ticket's attached JS test files... along with some 3D matrix tests I have. Looks good!
https://jira.appcelerator.org/browse/TIMOB-16789

jquick-axway reviewed (approved):
CR: Pass

Commit:
Merge branch 'master' into animate-update-props

ssekhri commented:
FR Passed.
Verified on:
Mac OS: 10.15.4
SDK: 9.1.0.v20200430133928
Appc CLI: 8.0.0
JDK: 11.0.6
Node: 10.17.0
Studio: 6.0.0.202005141803
Xcode: 11.4
Device: iPhone X(v13.2), iPhone 11 simulator (v13.4), iPad Pro 12.9 inch simulator (v13.4)
